### PR TITLE
Update pip window when largest video changes

### PR DESCRIPTION
**Pull request by beaufortfrancois**
After PiP window has been created, monitor every second if largest playing video changed so that video conferencing web apps always show the most appropriate video.
  
@mounirlamouri PTAL

### CODE REVIEW

**Comment by mounirlamouri** [comment]
Could we use the ResizeObserver API when available in order to avoid polling?

**Comment by beaufortfrancois** [comment]
This API is only useful for existing videos. We would still need to poll for new videos sadly

**Comment by mounirlamouri** [comment]
The way I see it, there is only one largest video, if we assume that the main video has to be resized in order for it to not be the main video, wouldn't that resize be caught by the ResizeObserver and at that point, we could run the algorithm to check which video is now the main video?

**Comment by beaufortfrancois** [comment]
Your observation is correct. https://github.com/GoogleChromeLabs/picture-in-picture-chrome-extension/pull/42/commits/d250e99a3693f360642e93d2edcecd706570c5b0 addresses it. Let me know what you think.

File: src/script.js [hunk 1]
```diff
@@ -22,19 +22,42 @@
...
-      video.removeAttribute('__pip__');
-    }, { once: true });
-    chrome.runtime.sendMessage({ message: 'enter' });
+    document.exitPictureInPicture();
+    return;
   }
+  await requestPictureInPicture(video);
+  chrome.runtime.sendMessage({ message: 'enter' });
+
+  new ResizeObserver(maybeUpdatePictureInPictureVideo).observe(video);
```
Comment (mounirlamouri) [inline src/script.js]: Shouldn't we update the video we observer when `maybeUpdatePictureInPictureVideo` is called? See suggestion above.

File: src/script.js [hunk 2]
```diff
@@ -22,19 +22,42 @@
...
+  }, { once: true });
+}
+
+async function maybeUpdatePictureInPictureVideo() {
+  if (!document.querySelector('[__pip__]')) {
+    return;
+  }
+  const video = findLargestPlayingVideo();
+  if (video && !video.hasAttribute('__pip__')) {
+    await requestPictureInPicture(video);
```
Comment (mounirlamouri) [inline src/script.js]: Shouldn't this be:
```
  await requestPictureInPicture(video);
  resizeObserver.unobserve(current_pip_video);
  resizeObserver.observe(video);
```

It implies that we save the value returned by the `querySelector()` call above.

File: src/script.js (hunk 1 above)
Comment (mounirlamouri) [reply src/script.js]: We should probably save the ResizeObserver object too, right?

File: src/script.js (hunk 2 above)
Comment (beaufortfrancois) [reply src/script.js]: How would you save the value of a video DOM element in the context of an extension script?

Comment (mounirlamouri) [reply src/script.js]: The video would be using `document.querySelector('[__pip__]')` but I guess your point applies for the `ResizeObserver`. I think we could use the fact that the `ResizeObserverCallback` includes the `ResizeObserver` as a second argument and we could `unobserver()` `document.querySelector('[__pip__]]')` and instead observer the new video. This way we avoid observing all possible videos but only the current main video. WDYT? I may try this later today...

Comment (beaufortfrancois) [reply src/script.js]: I didn't realize the second argument of ResizeObserverCallback was observer. Thank you really much! See last commit: https://github.com/GoogleChromeLabs/picture-in-picture-chrome-extension/pull/42/commits/34970f6a40a5c516b7515736d7cc7357e0c8553a

**Pull request merged**